src/main/java/com/contrastsecurity/cassandra/migration/CassandraMigration.java

```diff
@@ -132,38 +140,46 @@ <T> T execute(Action<T> action) {
         T result;
 
         VersionPrinter.printVersion(classLoader);
-
-        com.datastax.driver.core.Cluster cluster = null;
-        Session session = null;
         try {
             if (null == keyspace)
                 throw new IllegalArgumentException("Unable to establish Cassandra session. Keyspace is not configured.");
 
-            if (null == keyspace.getCluster())
-                throw new IllegalArgumentException("Unable to establish Cassandra session. Cluster is not configured.");
-
-            com.datastax.driver.core.Cluster.Builder builder = new com.datastax.driver.core.Cluster.Builder();
-            builder.addContactPoints(keyspace.getCluster().getContactpoints()).withPort(keyspace.getCluster().getPort());
-            if (null != keyspace.getCluster().getUsername() && !keyspace.getCluster().getUsername().trim().isEmpty()) {
-                if (null != keyspace.getCluster().getPassword() && !keyspace.getCluster().getPassword().trim().isEmpty()) {
-                    builder.withCredentials(keyspace.getCluster().getUsername(),
-                            keyspace.getCluster().getPassword());
-                } else {
-                    throw new IllegalArgumentException("Password must be provided with username.");
+
+            if (session == null) {
+
+                if (null == keyspace.getCluster())
+                    throw new IllegalArgumentException("Unable to establish Cassandra session. Cluster is not configured.");
+
+                com.datastax.driver.core.Cluster.Builder builder = new com.datastax.driver.core.Cluster.Builder();
```

Comment:
I'm performing an upgrage to cassandra 4 on my project too.
Do you have any reason to use com.datastax.driver.core.Cluster.Builder and not com.datastax.oss.driver.api.core.CqlSessionBuilder ?

Reply:
You are right. There were more changes I just checked in all those files. Pls. take a look.